File: frontend/src/api/datasets.ts
import api, { ApiResponse } from './index'

// ============== Types ==============

export interface DatasetRecord {
  id: number
  drug_id: string
  drug_name: string | null
  disease_id: string
  disease_name: string | null
  label: number
}

export interface Dataset {
  id: number
  name: string
  description: string | null
  source: string
  original_filename: string
  file_size: number
  drug_count: number
  disease_count: number
  association_count: number
  positive_count: number
  negative_count: number
  is_parsed: boolean
  parse_error: string | null
  created_by: number | null
  created_at: string
  updated_at: string
}

export interface DatasetStats {
  drug_count: number
  disease_count: number
  association_count: number
  positive_count: number
  negative_count: number
  positive_ratio: number
}

export interface DatasetListResponse {
  items: Dataset[]
  total: number
  page: number
  page_size: number
  pages: number
}

export interface DatasetPreviewResponse {
  records: DatasetRecord[]
  total: number
  page: number
  page_size: number
  pages: number
  columns: string[]
}

export interface DatasetFilter {
  keyword?: string
  source?: string
  start_date?: string
  end_date?: string
}

export interface DatasetUpdateRequest {
  name?: string
  description?: string
  source?: string
}

// ============== API Functions ==============

/**
 * Upload a new dataset
 */
export async function uploadDataset(
  file: File,
  name: string,
  source: string = 'custom',
  description?: string
): Promise<ApiResponse<Dataset>> {
  const formData = new FormData()
  formData.append('file', file)
  formData.append('name', name)
  formData.append('source', source)
  if (description) {
    formData.append('description', description)
  }

  return api.post('/datasets', formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  })
}

/**
 * List datasets with filtering and pagination
 */
export async function listDatasets(
  page: number = 1,
  pageSize: number = 10,
  filters?: DatasetFilter
): Promise<ApiResponse<DatasetListResponse>> {
  const params: Record<string, string | number> = {
    page,
    page_size: pageSize,
  }

  if (filters?.keyword) {
    params.keyword = filters.keyword
  }
  if (filters?.source) {
    params.source = filters.source
  }
  if (filters?.start_date) {
    params.start_date = filters.start_date
  }
  if (filters?.end_date) {
    params.end_date = filters.end_date
  }

  return api.get('/datasets', { params })
}

/**
 * Get dataset by ID
 */
export async function getDataset(id: number): Promise<ApiResponse<Dataset>> {
  return api.get(`/datasets/${id}`)
}

/**
 * Get dataset preview (paginated records)
 */
export async function getDatasetPreview(
  id: number,
  page: number = 1,
  pageSize: number = 50
): Promise<ApiResponse<DatasetPreviewResponse>> {
  return api.get(`/datasets/${id}/preview`, {
    params: { page, page_size: pageSize },
  })
}

/**
 * Get dataset statistics
 */
export async function getDatasetStats(id: number): Promise<ApiResponse<DatasetStats>> {
  return api.get(`/datasets/${id}/stats`)
}

/**
 * Update dataset metadata
 */
export async function updateDataset(
  id: number,
  data: DatasetUpdateRequest
): Promise<ApiResponse<Dataset>> {
  return api.put(`/datasets/${id}`, data)
}

/**
 * Delete a dataset
 */
export async function deleteDataset(id: number): Promise<ApiResponse<null>> {
  return api.delete(`/datasets/${id}`)
}

// Default export
export default {
  uploadDataset,
  listDatasets,
  getDataset,
  getDatasetPreview,
  getDatasetStats,
  updateDataset,
  deleteDataset,
}

File: backend/app/api/datasets.py
"""Dataset API endpoints."""
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, File, Form, Query, UploadFile
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import get_current_user, get_optional_user
from app.models.user import User
from app.schemas.dataset import (
    DatasetResponse,
    DatasetListResponse,
    DatasetPreviewResponse,
    DatasetStats,
    DatasetUpdate,
)
from app.schemas.response import ResponseModel
from app.services.dataset import DatasetService


router = APIRouter(prefix="/datasets", tags=["Datasets"])


@router.post("", response_model=ResponseModel[DatasetResponse])
async def upload_dataset(
    name: str = Form(..., min_length=1, max_length=200),
    source: str = Form(default="custom", max_length=50),
    description: Optional[str] = Form(default=None),
    file: UploadFile = File(...),
    db: AsyncSession = Depends(get_db),
    current_user: Optional[User] = Depends(get_optional_user),
):
    """
    Upload a new dataset.
    
    The file should be a CSV or TSV file with the following columns:
    - drug_id (required): Drug identifier
    - drug_name (optional): Drug name
    - disease_id (required): Disease identifier
    - disease_name (optional): Disease name
    - label (optional): Association label (0 or 1, defaults to 1)
    """
    service = DatasetService(db)
    user_id = current_user.id if current_user else None
    
    dataset = await service.create_dataset(
        name=name,
        source=source,
        description=description,
        file=file,
        user_id=user_id,
    )
    
    return ResponseModel(
        code=200,
        message="Dataset uploaded successfully",
        data=DatasetResponse.model_validate(dataset),
    )


@router.get("", response_model=ResponseModel[DatasetListResponse])
async def list_datasets(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=10, ge=1, le=100),
    keyword: Optional[str] = Query(default=None),
    source: Optional[str] = Query(default=None),
    start_date: Optional[datetime] = Query(default=None),
    end_date: Optional[datetime] = Query(default=None),
    db: AsyncSession = Depends(get_db),
):
    """List datasets with optional filtering and pagination."""
    service = DatasetService(db)
    
    result = await service.list_datasets(
        page=page,
        page_size=page_size,
        keyword=keyword,
        source=source,
        start_date=start_date,
        end_date=end_date,
    )
    
    return ResponseModel(
        code=200,
        message="Success",
        data=result,
    )


@router.get("/{dataset_id}", response_model=ResponseModel[DatasetResponse])
async def get_dataset(
    dataset_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Get dataset details by ID."""
    service = DatasetService(db)
    dataset = await service.get_dataset_or_404(dataset_id)
    
    return ResponseModel(
        code=200,
        message="Success",
        data=DatasetResponse.model_validate(dataset),
    )


@router.get("/{dataset_id}/preview", response_model=ResponseModel[DatasetPreviewResponse])
async def get_dataset_preview(
    dataset_id: int,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=50, ge=1, le=500),
    db: AsyncSession = Depends(get_db),
):
    """Get paginated preview of dataset records."""
    service = DatasetService(db)
    result = await service.get_dataset_preview(
        dataset_id=dataset_id,
        page=page,
        page_size=page_size,
    )
    
    return ResponseModel(
        code=200,
        message="Success",
        data=result,
    )


@router.get("/{dataset_id}/stats", response_model=ResponseModel[DatasetStats])
async def get_dataset_stats(
    dataset_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Get dataset statistics."""
    service = DatasetService(db)
    stats = await service.get_dataset_stats(dataset_id)
    
    return ResponseModel(
        code=200,
        message="Success",
        data=stats,
    )


@router.put("/{dataset_id}", response_model=ResponseModel[DatasetResponse])
async def update_dataset(
    dataset_id: int,
    update_data: DatasetUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Update dataset metadata."""
    service = DatasetService(db)
    dataset = await service.update_dataset(dataset_id, update_data)
    
    return ResponseModel(
        code=200,
        message="Dataset updated successfully",
        data=DatasetResponse.model_validate(dataset),
    )


@router.delete("/{dataset_id}", response_model=ResponseModel)
async def delete_dataset(
    dataset_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Delete a dataset."""
    service = DatasetService(db)
    await service.delete_dataset(dataset_id)
    
    return ResponseModel(
        code=200,
        message="Dataset deleted successfully",
    )
